research/src/pages/select/select.proposal.mdx

```diff
@@ -201,7 +202,47 @@ and is therefore considered to be invalid.
   - Presses the escape key
   - Selects one, or more (if `multiple` is true), by clicking or hitting `enter` or `space` key
 
-### Keyboard traversal
+### Typeahead
+
+#### The <a class="link" href="#typeahead-steps" id="typeahead-steps">typeahead steps</a> given a `<select>` and a `typed character` are as follows:
+
+1. Let `start new search` be false.
+1. If longer than the <a href="#typeahead-search-timeout">typeahead search timeout</a>
+   has elapsed since the previous invocation of the [typeahead steps](#typeahead-steps)
+   for `listbox`, then set `start new search` to true and set `buffer` to an empty string.
+1. Let `buffer` be `<select>`'s [`typeahead buffer`](#typeahead-buffer)
+1. Append `typed character` to `buffer`.
+1. Let `options` be the `<select>`'s list of options.
+1. Let `focusedIndex` be the index of `<select>`'s currently focused option.
+1. If `start new search` is true, then let `options for search` be the array of options given by
+   `Array.prototype.slice.call(options, focusedIndex + 1, options.length).concat(Array.prototype.slice.call(options, 0, focusedIndex + 1))`.
+   Otherwise, let `options for search` be the array of options given by
+   `Array.prototype.slice.call(options, focusedIndex, options.length).concat(Array.prototype.slice.call(options, 0, focusedIndex))`.
+   <p class="note">
+     The point is that if we're in the middle of a search, we should keep trying to match on the
+     currently selected &lt;option&gt;. If this is a new search, we should start searching at the
+     &lt;option&gt; after the currently selected &lt;option&gt;. If the user wanted the current
+     &lt;option&gt;, then presumably they wouldn't still be typing search text.
+   </p>
+1. Let `newly focused option` be the result of [finding the first matching
+   option](#find-the-first-matching-option) given `buffer` and `options for search`.
+1. If `newly focused option` is not null, move focus to `newly focused option`.
+
+#### To <a class="link" href="#find-the-first-matching-option" id="find-the-first-matching-option">find the first matching option</a> given a character buffer `buffer` and a list of options `options`, run these steps:
+
+1. For each `<option>` `option` in `options`:
+   1. If `buffer` is a case-insensitive prefix of `option.label`, then then return `option`.
+1. Return null.
+
+#### The <a class="link" href="#typeahead-buffer" id="typeahead-buffer">`typeahead buffer`</a> of a `<select>`
+
+A buffer containing the characters typed by the user for typeahead. Initially an empty string.
+
+#### <a class="link" href="#typeahead-search-timeout" id="typeahead-search-timeout">Typeahead search timeout</a>
+
+This is the maximum time that may elapse between keystrokes without the typeahead
```

Comment:
Given that we want spaces to work within the typeahead should we remove that listener at the start of the algo and then add it back in during this section?

Reply:
'Space' will also trigger the 'typeahead steps' (after I addressed your feedback that the character for the event should not just be alphanumeric), so I don't think it should require extra handling with respect to the timeout.

Reply:
I'm confused how this will be the case given that the space event will still be tied to the listbox to open it unless I'm misunderstanding how the listeners are working and effectively the space event will be blocked while the method is running. Even if that is the case though, we should be clear in what can be leveraged in type ahead that even though space won't trigger the typeahead method it should be usable within the typeahead algo. Looking at the algo it doesn't seem to limit any chars but space is one that selects an option. I'm not sure how to reconcile those two

Reply:
Sorry, I misunderstood, I see what you're saying now.

I'd actually overlooked that we had `keydown(space)` handlers in `listbox`.  For Edge/Chrome and Firefox, 'space' doesn't do anything when the popup is opened.  Did we diverge from that behavior deliberately?  I'm tempted to say that we should converge with the current browser behavior and remove the `keydown(space)` handlers from the `listbox` (the ones on `button` would still be there, so that 'space' could still be used to open the popup). No functionality would be lost because it 'space' currently has the same behavior as 'Enter' for listbox.  Looking at the design systems listed at https://open-ui.org/components/select.research#names, (looking at button/listbox selects only, not the comboboxes with text input whose typeahead will work differently) I found these results:
Space doesn't do anything: [Ant Design](https://ant.design/components/select/), [Evergreen](https://evergreen.segment.com/components/select/), [Lightning](https://www.lightningdesignsystem.com/components/select/) (this one has typeahead), [Semantic UI](https://semantic-ui.com/modules/dropdown.html#selection), [Bootstrap](https://getbootstrap.com/docs/4.3/components/dropdowns/) (space toggles when on the button, but not when focus tabs into the listbox)
Space closes the popup, no typeahead: [Atlassian](https://atlassian.design/components/select/examples#popup-select), [Material](https://material-components.github.io/material-components-web-catalog/#/component/select), [Fluent UI](https://developer.microsoft.com/en-us/fluentui#/controls/web/dropdown)
Space closes the popup, has typeahead: [Carbon](https://www.carbondesignsystem.com/components/select/code/)
Carbon is funny because in their [live demo](https://www.carbondesignsystem.com/components/select/code/#live-demo) you can see that closing on space breaks their typahead.  Try selecting "Gray 100" via typeahead -- it's painful.

If we want to keep the behavior where 'space' closes the listbox and/or selects an option, I only see two possibilities:
1. Don't allow 'space' in typeahead.  This means that multi-word examples can't be easily selected via typahead, as in the Carbon example.
1. Treat 'space' as part of the typeahead if a search is underway, but make it select an option/close the popup otherwise.  This seems even more confusing and opaque to users than the current typeahead behavior in browsers.

Neither of these are great so I'm in favor of removing the `keyboard(space)` handlers from the listbox and relying on Enter for that functionality.

Reply:
ah, interesting - good catch. I'm so used to space invoking controls that I must have assumed it would select and change state.

Reply:
I filed an issue for this here: https://github.com/WICG/open-ui/issues/171